client/extensions/woocommerce/state/sites/products/reducer.js

```diff
@@ -65,6 +67,15 @@ export function productsRequestSuccess( state, action ) {
 	};
 }
 
+export function productsDeleteSuccess( state, action ) {
+	const prevState = state || {};
+	const prevProducts = prevState.products || [];
+	const newProducts = prevProducts.filter( p => p.id !== action.data.id );
```

Comment:
This could be simplified a little by using lodash reject instead, e.g.

reject( prevProducts, { 'id': action.data.id } )